src/libcore/mem/manually_drop.rs

```diff
@@ -7,14 +7,20 @@ use crate::ptr;
 ///
 /// `ManuallyDrop<T>` is subject to the same layout optimizations as `T`.
 /// As a consequence, it has *no effect* on the assumptions that the compiler makes
-/// about all values being initialized at their type.  In particular, initializing
-/// a `ManuallyDrop<&mut T>` with [`mem::zeroed`] is undefined behavior.
+/// about its contents: the validity rules for `ManuallyDrop<T>` are the same as those
+/// for `T`. For example, initializing a `ManuallyDrop<&mut T>` with [`mem::zeroed`]
+/// is undefined behavior.
 /// If you need to handle uninitialized data, use [`MaybeUninit<T>`] instead.
 ///
+/// `ManuallyDrop<T>` *does* allow violating invariants normally enforced by the type
+/// system, and for that reason many of its functions are unsafe. For example, the
+/// `ManuallyDrop::drop` method could be used to drop a value multiple times, which
+/// is prohibited in safe code.
```

Comment:
Surprisingly many of its functions are safe though! In particular, `deref`/`deref_mut`. This demonstrates that indeed the safety invariant of `ManuallyDrop<T>` is the same as that of `T`. (As is the validity invariant.)

So, when `ManuallyDrop::drop` returns, *passing that value to arbitrary safe code is unsound*.

I am not entirely sure what the point of this paragraph is.

Reply:
Hm, I was trying to capture what made `ManuallyDrop` special so it would be easier to compare to MaybeUninit, but I see what you mean.

I suppose the only thing this actually enables that wasn't possible before is "forgetting a value in-place", and this is fine as long as the value isn't pinned.

Reply:
> Hm, I was trying to capture what made ManuallyDrop special so it would be easier to compare to MaybeUninit,

In terms of my experimental terminology -- `MaybeUninit` has a trivial validity and safety invariant; `ManuallyDrop` inherits both of them from `T`.

But I am not sure what the best way is to explain that...